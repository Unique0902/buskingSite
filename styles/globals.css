@tailwind base;
@tailwind components;
@tailwind utilities;
/* TODO: 글꼴 불러오는 로딩 줄이거나 없애기 */
* {
  box-sizing: border-box;
  font-family: 'Pretendard';
}

@font-face {
  font-family: 'Pretendard';
  font-weight: 45 920;
  font-style: normal;
  font-display: swap;
  src: url('../assets/fonts/PretendardVariable.woff2')
    format('woff2-variations');
}

li {
  list-style: none;
}
body {
  margin: 0;
  @apply text-black dark:text-white dark:bg-slate-600;
}
p,
h1,
h2 {
  margin: 0;
}
ul {
  padding: 0;
}
