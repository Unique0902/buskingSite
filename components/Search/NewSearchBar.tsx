import React, { createContext, ReactNode, useContext, useState } from 'react';
import { NewSearchWord } from '../../store/type/searchword';
import { xyPadding } from '../../styles/theme';
import PrimaryBtn from '../Btn/PrimaryBtn';

interface SearchBarProps {
  children: ReactNode;
  categories: string[];
}
type SearchBarContextProps = {
  searchWord: NewSearchWord;
  categories: string[];
  handleInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleSelectChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
};

const SearchBarContext = createContext<SearchBarContextProps>(
  {} as SearchBarContextProps
);

const NewSearchBar = ({ children, categories }: SearchBarProps) => {
  const [searchWord, setSearchWord] = useState<NewSearchWord>({
    name: '',
    category: categories[0],
  });
  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchWord({ ...searchWord, name: e.target.value });
  };
  const handleSelectChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    setSearchWord({
      ...searchWord,
      category: e.target.value,
    });
  };

  return (
    <SearchBarContext.Provider
      value={{ searchWord, categories, handleInputChange, handleSelectChange }}
    >
      <form
        className={`relative flex flex-row items-center justify-center flex-1 gap-2 mb-6 max-lg:w-full`}
        onSubmit={(e) => {
          e.preventDefault();
        }}
      >
        {children}
      </form>
    </SearchBarContext.Provider>
  );
};

type MainSecProps = {
  children: ReactNode;
};

const MainSec = ({ children }: MainSecProps) => {
  return (
    <div className='relative flex flex-row items-center justify-center flex-1 gap-2 max-lg:w-full max-lg:mx-2'>
      {children}
    </div>
  );
};

type SubSecProps = {
  children: ReactNode;
};
const SubSec = ({ children }: SubSecProps) => {
  return <>{children}</>;
};

const Select = () => {
  const { searchWord, categories, handleSelectChange } =
    useContext(SearchBarContext);
  return (
    <select
      className='p-2 text-lg border-2 border-black rounded-xl max-lg:mr-0 max-lg:text-base dark:bg-slate-500 '
      value={searchWord.category}
      onChange={handleSelectChange}
    >
      {categories.map((val) => (
        <option key={'selectOption' + val} value={val}>
          {val}
        </option>
      ))}
    </select>
  );
};

const Option = ({ value }: { value: string }) => {
  return <option value={value}>{value}</option>;
};

type InputProps = {
  placeholder?: string;
};

const Input = ({ placeholder = '검색어를 입력하세요..' }: InputProps) => {
  const { searchWord, handleInputChange } = useContext(SearchBarContext);
  return (
    <div className='relative flex flex-row items-center flex-1'>
      <input
        type='search'
        className='w-full p-2 text-lg border-2 border-black rounded-xl max-lg:text-base dark:bg-slate-500'
        placeholder={placeholder}
        value={searchWord.name}
        onChange={handleInputChange}
      />
    </div>
  );
};

type ButtonProps = {
  handleClickBtn: (searchWord: NewSearchWord) => void;
  text: string;
};

const Button = ({ handleClickBtn, text }: ButtonProps) => {
  const { searchWord } = useContext(SearchBarContext);
  return (
    <PrimaryBtn
      handleClick={() => handleClickBtn(searchWord)}
      btnPadding={xyPadding.base}
      isSubmit={true}
    >
      {text}
    </PrimaryBtn>
  );
};

NewSearchBar.MainSec = MainSec;
NewSearchBar.SubSec = SubSec;

MainSec.Select = Select;
MainSec.Option = Option;
MainSec.Input = Input;
MainSec.Button = Button;

export default NewSearchBar;
