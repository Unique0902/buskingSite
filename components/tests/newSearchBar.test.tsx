import { describe, expect, it, jest } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import NewSearchBar from '../Search/NewSearchBar';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';

describe('newSearchBar component tests', () => {
  const testCategories = ['category1', 'category2'];

  it('render test', () => {
    const component = renderer.create(
      <NewSearchBar categories={testCategories}>
        <NewSearchBar.MainSec>
          <NewSearchBar.MainSec.Select />
          <NewSearchBar.MainSec.Input />
          <NewSearchBar.MainSec.Button handleClickBtn={() => {}} text='btn' />
        </NewSearchBar.MainSec>
        <NewSearchBar.SubSec>
          <div>subsec</div>
        </NewSearchBar.SubSec>
      </NewSearchBar>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('categoriesArr change to each option', () => {
    render(
      <NewSearchBar categories={testCategories}>
        <NewSearchBar.MainSec>
          <NewSearchBar.MainSec.Select />
        </NewSearchBar.MainSec>
      </NewSearchBar>
    );
    const [category1, category2] = screen.getAllByRole('option');

    expect(category1.textContent).toBe('category1');
    expect(category2.textContent).toBe('category2');
  });

  it('when btn click, handleClickBtn execute with searchWord', async () => {
    const handleClickBtn = jest.fn();
    render(
      <NewSearchBar categories={testCategories}>
        <NewSearchBar.MainSec>
          <NewSearchBar.MainSec.Select />
          <NewSearchBar.MainSec.Input />
          <NewSearchBar.MainSec.Button
            handleClickBtn={handleClickBtn}
            text='btn'
          />
        </NewSearchBar.MainSec>
      </NewSearchBar>
    );
    const input = screen.getByRole('searchbox');
    await userEvent.type(input, 'text');
    expect(input).toHaveValue('text');
    const btn = screen.getByRole('button');
    await userEvent.click(btn);
    expect(handleClickBtn).toBeCalledWith({
      category: 'category1',
      name: 'text',
    });
  });
});
