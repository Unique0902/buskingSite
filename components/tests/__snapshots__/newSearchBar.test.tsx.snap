// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`newSearchBar component tests render test 1`] = `
<form
  className="relative flex flex-row items-center justify-center flex-1 gap-2 mb-6 max-lg:w-full"
  onSubmit={[Function]}
>
  <div
    className="relative flex flex-row items-center justify-center flex-1 gap-2 max-lg:w-full max-lg:mx-2"
  >
    <select
      className="p-2 text-lg border-2 border-black rounded-xl max-lg:mr-0 max-lg:text-base dark:bg-slate-500 "
      onChange={[Function]}
      value="category1"
    >
      <option
        value="category1"
      >
        category1
      </option>
      <option
        value="category2"
      >
        category2
      </option>
    </select>
    <div
      className="relative flex flex-row items-center flex-1"
    >
      <input
        className="w-full p-2 text-lg border-2 border-black rounded-xl max-lg:text-base dark:bg-slate-500"
        onChange={[Function]}
        placeholder="검색어를 입력하세요.."
        type="search"
        value=""
      />
    </div>
    <button
      className="hover:opacity-70 cursor-pointer rounded-lg"
      onClick={[Function]}
      style={
        {
          "backgroundColor": "#0055ff",
          "color": "#FFFFFF",
          "fontSize": "12px",
          "padding": "12px 24px",
        }
      }
      type="submit"
    >
      btn
    </button>
  </div>
  <div>
    subsec
  </div>
</form>
`;
